### PR TITLE
Shamir deletion specification - server side

### LINKED ISSUE #7364: [Shamir port] Certificate deletion
Related #6090 

When a shamir setup is overwritten/deleted a deletion certificate must be produced and sent to the author and the share recipients.

### PR DESCRIPTION
**Pull request by AureliaDolo**
Related #7364

### CODE REVIEW

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [inline docs/rfcs/1000-shamir-based-recovery.md:843]:
```suggestion
| brief       | x      | x               |
```

The shamir_recovery_others_list provides both certificates. I'm not exactly sure why though

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (vxgmichel) [inline docs/rfcs/1000-shamir-based-recovery.md:764]: We tend to use the synonym `inconsistent` in the code base:
```
± git grep inconsistent | wc -l
15
[vinmic:~/repos/parsec-cloud-v3] [base] master+*
± git grep incoherent | wc -l  
4
```

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (vxgmichel) [inline docs/rfcs/1000-shamir-based-recovery.md:829]: There is no "share recipients mismatch" status in `shamir_recovery_deletion`, is that right?

Comment (vxgmichel) [inline docs/rfcs/1000-shamir-based-recovery.md:829]: Maybe acknowledge that `previous_timestamp` and `share_recipients` were not strictly required here, as this information is already available by browsing the shamir topic. If we do include this information, it'd be good to have some kind of justification for it (clarity, or something else, I don't remember what the conclusions were about this topic).

**Review by vxgmichel — APPROVED** [review]
A couple of comments but LGTM otherwise :+1: 

@touilleMan should review and validate this before it's merged

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (vxgmichel) [reply docs/rfcs/1000-shamir-based-recovery.md:764]: The 4 occurrences of `incoherent` were from your previous PR, sorry for not catching that earlier.

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:829]: Yes that was what I mean by the line below ^^

It was as a mean to link to the previous certificate. But I saw that @touilleMan is working on a better solution here #7577

Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:829]: Yes indeed, we can remove that.

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:764]: Nice catch :fishing_pole_and_fish: 

So here you have the correction #7621

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [inline docs/rfcs/1000-shamir-based-recovery.md:764]:
```suggestion
                "status": "shamir_setup_not_found"
```

to stay consistent with e.g. [`user_not_found`](https://github.com/Scille/parsec-cloud/blob/048b4c6528cbb1b78b6daa9d567f356c6307462f/libparsec/crates/protocol/schema/authenticated_cmds/user_revoke.json5#L23) and [`realm_not_found`](https://github.com/Scille/parsec-cloud/blob/048b4c6528cbb1b78b6daa9d567f356c6307462f/libparsec/crates/protocol/schema/authenticated_cmds/realm_share.json5#L33
And `shamir_setup` instead of just `shamir` given it's how it is specified in `shamir_recovery_setup`:
https://github.com/Scille/parsec-cloud/blob/048b4c6528cbb1b78b6daa9d567f356c6307462f/libparsec/crates/protocol/schema/authenticated_cmds/shamir_recovery_setup.json5#L54

I'm also okay to stick with just `setup`, so the vocabulary would be "a `ShamirRecoveryBrief` describes a setup, and has been added in the server with the `shamir_recovery_setup` command", but in this case we have to correct the `shamir_recovery_setup` command and make sure we are consistent everywhere (for instance I didn't check the comments in the code)

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [inline docs/rfcs/1000-shamir-based-recovery.md:741]:
```suggestion
            "cmd": "shamir_recovery_delete_setup",
```

The command name should contain a verb

(btw I guess `shamir_recovery_setup` is a bit weird on this regard given `setup` is both the verb and the noun, since after that we refer to the configuration set by this command as the `shamir setup`)

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:825]: After thinking about it a bit more, I'd rather keep the share recipients in the certificate, even if we could retrieve the information, because it's a certificate we went them to get.

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [reply docs/rfcs/1000-shamir-based-recovery.md:231]: On top of that, the `shamir_setup_already_exists` should have a `DateTime` field `last_shamir_certificate_timestamp`.

This is important given the Parsec client is expected to eagerly download all certificates, hence it should have been aware of the already existing shamir setup. And if that's not the case it is because the client is lagging behind (typically the certificate is very new), and in this case the client polls the server for certificates... but there is great chances this is already under way (the certificates ops start polling as soon as it gets the "new certificates" events from the server), hence this `last_shamir_certificate_timestamp` that allows to detect if the poll is really needed or if it has been concurrently achieved.

See for instance what is done for `user_revoke` command:
https://github.com/Scille/parsec-cloud/blob/048b4c6528cbb1b78b6daa9d567f356c6307462f/libparsec/crates/protocol/schema/authenticated_cmds/user_revoke.json5#L26-L31
and in the client (note the `RemoteIdempotent` status that means the server rejected our certificate, but we are missing some certificates and once we fetch them we will discover our current operation has already been achieved):
https://github.com/Scille/parsec-cloud/blob/048b4c6528cbb1b78b6daa9d567f356c6307462f/libparsec/crates/client/src/client/user_revoke.rs#L15-L26

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [inline docs/rfcs/1000-shamir-based-recovery.md:760]: I would merge `incoherent_previous_timestamp` into `no_setup_to_delete`: having both error makes sens if we consider the server first look for the only certificate that can be deleted, then check that its timestamp is what the client has sent.

However another way to see things is to consider the timestamp is used to uniquely identify a setup to delete (see another suggestion I did below, where I suggest to use user ID + timestamp as unique identifier of shamir setup). In this case if the unique identifier leads to nothing it's normal to return a `setup_not_found` error.

What makes me feel the balance tips toward the second solution is this comment: `// Future evolution 1: Shamir recovery has already been setup, should ask your admin to reset it first !`. So it seems interesting to later allow admin to remove other people's shamir setup. With the "user ID + timestamp is the unique identifier of the setup" there is nothing to change to the API to support this usecase.

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [reply docs/rfcs/1000-shamir-based-recovery.md:764]: On top of that I would add a `shamir_setup_already_deleted` (with a `DateTime` field `last_shamir_recovery_certificate`) to cover the case the shamir setup exists but can no longer be deleted (see the comment about `shamir_setup_already_exists` for the reason for the field `last_shamir_recovery_certificate`)

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (touilleMan) [inline docs/rfcs/1000-shamir-based-recovery.md:833]: This is no longer needed if we have the user ID + timestamp (and is complex to add later on anyway given we have to support backward compatibility and hence not all certificates will have the ID...)

So I would suggest to just remove this comment

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:231]: > On top of that, the shamir_setup_already_exists should have a DateTime field last_shamir_certificate_timestamp. 

Done in #7622 :tada:

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (vxgmichel) [reply docs/rfcs/1000-shamir-based-recovery.md:825]: Sorry for not being clear, I meant that we probably need to add a "share recipients mismatch" status to `shamir_recovery_deletion` to return when this field doesn't match what is in the previous certificate.

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md:820]: >     // (i.e. a user can only remove its own Shamir recovery)

And that means if we want a configuration where only admins can delete shamir certificates, it should be quite easy

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [reply docs/rfcs/1000-shamir-based-recovery.md]: Maybe `shamir_recovery_setup_setup` ?

File: docs/rfcs/1000-shamir-based-recovery.md
Comment (AureliaDolo) [inline docs/rfcs/1000-shamir-based-recovery.md:761]: Share recipients are the only not identifying data that we have in the deletion certificate to check. Maybe we could include more data from the brief to ensure coherence. I'm not sure how much it's needed though

**Review by touilleMan — APPROVED** [review]

**Pull request merged**
